lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
 	DeleteToken(ctx context.Context, token string) error
 }
 
+type NodeJoinWait struct {
+	cfg *NodeJoinWaitConfig
+}
+
+func NewNodeJoinWait(cfg *NodeJoinWaitConfig) (*NodeJoinWait, error) {
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	return &NodeJoinWait{cfg: cfg}, nil
+}
+
+// Run grabs the host UUID of an agent from disk and then waits for the node with the given name to
+// show up in the cluster.
+//
+// The Electron app calls this method soon after starting the agent process.
+func (n *NodeJoinWait) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, cluster *clusters.Cluster) (clusters.Server, error) {
+	nodeName, err := n.getNodeNameFromHostUUIDFile(ctx, cluster)
+	if err != nil {
+		return clusters.Server{}, err
+	}
+
+	server, err := n.waitForNode(ctx, accessAndIdentity, cluster, nodeName)
+	if err != nil {
+		return clusters.Server{}, trace.Wrap(err)
+	}
+
+	return clusters.Server{
+		URI:    cluster.URI.AppendServer(server.GetName()),
+		Server: server,
+	}, nil
+}
+
+func (n *NodeJoinWait) getNodeNameFromHostUUIDFile(ctx context.Context, cluster *clusters.Cluster) (string, error) {
+	hostUUIDPath := filepath.Join(n.cfg.AgentsDir, cluster.ProfileName, "data", utils.HostUUIDFile)
+
+	// NodeJoinWait gets executed when the agent is booting up, so the host UUID file might not exist
+	// on disk yet. Use a ticker to periodically check for its existence.
+	ticker := time.NewTicker(50 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			out, err := utils.ReadPath(hostUUIDPath)
+			if err != nil {
+				if trace.IsNotFound(err) {
+					continue
+				}
+				return "", trace.Wrap(err)
+			}
+
+			id := strings.TrimSpace(string(out))
+			if id == "" {
+				return "", trace.NotFound("host UUID is empty")
+			}
+
+			return id, nil
+		case <-ctx.Done():
+			return "", trace.Wrap(ctx.Err(), "waiting for host UUID file to be created")
+		}
+	}
+}
+
+func (n *NodeJoinWait) waitForNode(ctx context.Context, accessAndIdentity AccessAndIdentity,
+	cluster *clusters.Cluster, nodeName string) (types.Server, error) {
+	watcher, err := initializeWatcher(ctx, accessAndIdentity, types.KindNode)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer watcher.Close()
+
+	// Attempt to fetch the node from the cluster manually, in case it's joined the cluster before we
+	// started the watcher.
+	//
+	// This means that we might return immediately if the node is still in the cache, even if
+	// technically the agent has not joined the cluster yet. We're fine with this edge case.
+	server, err := accessAndIdentity.GetNode(ctx, defaults.Namespace, nodeName)
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err)
+		}
+		// Continue in case of NotFound error.
+	} else {
+		return server, nil
+	}
+
+	resource, err := waitForOpPut(ctx, watcher, types.KindNode, nodeName)
```

Comment:
Wrong commit to comment, but I'm not seeing where we set a timeout for `waitForOpPut`.